keps/sig-cli/2906-kustomize-function-catalog/README.md

```diff
@@ -0,0 +1,667 @@
+
+In addition to container based plugin providers, the `Catalog` will support discovery of Starlark and Exec based plugin providers, via an HTTP(s), Git, or OCI reference as illustrated below: 
+
+```yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-plugins"
+spec: 
+  modules: 
+  - apiVersion: example.com/v1
+    kind: GroovyApplication
+    description: "A Kustomize plugin provider that can handle groovy apps"
+    provider:  
+      starlark: https://example.co/module_providers/starlark-func:v1.0.0
+```
+
+This concept can be extended later to support additional plugin provider packaging, but is out of scope for the current proposal.
+
+When HTTP(s) references are used, the HTTP(s) endpoint must support anonymous access for reading resources. Resources will be expected to be stored as a single file, such as `catalog.yaml`. Git or OCI references can be authenticated or anonymous, and will use appropriate configuration from the users file-system. 
+
+ 
+### User Stories (Optional)
+
+#### Story 1
+
+As a platform developer at enterprise company Example Co, I want to publish a catalog of Kustomize plugin providers that represent custom capabilities important for our internal Kubernetes platform. I have built and published several Kustomize plugins, packaged as Docker images, to our internal Docker registry: `docker.example.co`.
+
+To do this, I build a new `Catalog` API resource: 
+
+```yaml
+# catalog.yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-plugins"
+spec: 
+  modules: 
+  - apiVersion: example.com/v1
+    kind: JavaApplication
+    description: "A Kustomize plugin provider that can handle Java apps"
+    provider: 
+      container: 
+        image: docker.example.co/plugins/java:v1.0.0
+  - apiVersion: example.com/v1
+    kind: Logger
+    description: "A Kustomize plugin provider adds our bespoke logging"
+    provider: 
+      container: 
+        image: docker.example.co/plugins/logger:v1.0.0
+  - apiVersion: example.com/v1
+    kind: SecretSidecar
+    description: "A Kustomize plugin provider adds our bespoke secret sidecar"
+    provider: 
+      container: 
+        image: docker.example.co/plugins/secrets:v1.0.0
+```
+
+I then publish this catalog to https://example.co/kustomize/catalog.yaml, for use by Example Co Kustomize users.
+
+#### Story 2
+
+As an application developer at Example Co, I want to use the published Example Co Catalog in the `Kustomization` for my application, after locating the published location of the catalog.
+
+While building my `Kustomization`, I don't want to care about the provider configuration and I want Kustomize to figure things out for me, based on the catalog.
+
+
+```yaml
+# app/kustomization.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+catalogs:
+  - https://example.co/kustomize/catalog.yaml
+resources:
+- java.yaml
```

Comment:
Yes, but why use `resources`, rather than `generators` and `transformers` field which already exist specifically for this purpose?

If the function invocation resource of kind `JavaApplication` in `java.yaml` is put under `resources`, should the function also act on its own config defined on `java.yaml` as well? Should this `JavaApplication` resource be included in the `kustomize build` output? Putting it in the same place as all the user's other resources seems very confusing to me.

Reply:
Yeah, this is a bad example, will replace with transformers :)